File: 2016/app/src/main/java/aoc2016/solutions/Day12.java
package aoc2016.solutions;

import java.util.HashMap;

import aoc2016.utils.FileLoader;

public class Day12 {
    public static void solve() {
        String input = FileLoader.loadFile("day12.txt");
        System.out.println("Day 12 part 1: " + part1(input));
        System.out.println("Day 12 part 2: " + part2(input));
    }

    private static int solve(String input, HashMap<Character, Integer> registers) {
        String[] instructions = input.split("\n");
        int i = 0;
        while (i < instructions.length) {
            String[] tokens = instructions[i].split(" ");
            String action = tokens[0];
            switch (action) {
                case "cpy":
                    if (tokens[1].matches("[a-d]")) {
                        registers.put(tokens[2].charAt(0), registers.get(tokens[1].charAt(0)));
                    } else {
                        registers.put(tokens[2].charAt(0), Integer.parseInt(tokens[1]));
                    }
                    break;
                case "inc":
                    registers.put(tokens[1].charAt(0), registers.get(tokens[1].charAt(0)) + 1);
                    break;
                case "dec":
                    registers.put(tokens[1].charAt(0), registers.get(tokens[1].charAt(0)) - 1);
                    break;
                case "jnz":
                    if (tokens[1].matches("[a-d]")) {
                        if (registers.get(tokens[1].charAt(0)) != 0) {
                            i += Integer.parseInt(tokens[2]);
                            continue;
                        }
                    } else {
                        if (Integer.parseInt(tokens[1]) != 0) {
                            i += Integer.parseInt(tokens[2]);
                            continue;
                        }
                    }
                    break;
                default:
                    break;
            }
            i++;
        }
        return registers.get('a');
    }

    public static int part1(String input) {
        HashMap<Character, Integer> registers = new HashMap<>();
        registers.put('a', 0);
        registers.put('b', 0);
        registers.put('c', 0);
        registers.put('d', 0);
        return solve(input, registers);
    }

    public static int part2(String input) {
        HashMap<Character, Integer> registers = new HashMap<>();
        registers.put('a', 0);
        registers.put('b', 0);
        registers.put('c', 1);
        registers.put('d', 0);
        return solve(input, registers);
    }
}

File: 2017/app/src/main/kotlin/aoc/solutions/Day12.kt
package aoc.solutions

import aoc.utils.FileLoader
import java.util.*

fun solveDayTwelve() {
    val input = FileLoader("day12.txt").readLines()
    println("Day 12 part 1: ${partOne(input)}")
    println("Day 12 part 2: ${partTwo(input)}")
}

private fun partOne(input: List<String>): Int {
    val nodes = buildGraph(input)
    // From every node, dfs to 0
    return nodes.keys.map { dfs(it, nodes, true).contains("0") }.count { it }
}

private fun partTwo(input: List<String>): Int {
    val nodes = buildGraph(input)
    return nodes.keys.map {
        dfs(it, nodes).sorted()
    }.distinct().count()
}

private fun buildGraph(input: List<String>): HashMap<String, MutableList<String>> {
    val nodes = HashMap<String, MutableList<String>>()
    input.forEach {
        val parts = it.split(" <-> ")
        if (parts[0] !in nodes) {
            nodes[parts[0]] = mutableListOf()
        }
        for (child in parts[1].split(", ")) {
            if (child !in nodes) {
                nodes[child] = mutableListOf()
            }
            nodes[parts[0]]!!.add(child)
        }
    }

    return nodes
}

private fun dfs(
    start: String,
    nodes: HashMap<String, MutableList<String>>,
    containsZero: Boolean = false,
): List<String> {
    val stack = Stack<String>()
    val visited = HashSet<String>()
    val path = mutableListOf<String>()
    stack.push(start)
    visited.add(start)
    while (!stack.isEmpty()) {
        // Pop the stack
        val curr = stack.pop()
        // Add to path
        path.add(curr)
        if (containsZero && curr == "0") {
            return path
        }
        // Visit the adj nodes
        for (child in nodes[curr]!!) {
            if (child in visited) {
                continue
            }
            visited.add(child)
            stack.push(child)
        }
    }

    return path.toList()
}
